Implement Montgomery multiplication assembly optimization for RV64GC

fengpengboa opened the pull request:
This PR introduces the RV64GC-optimized assembly implementation for Montgomery multiplication, and significantly accelerating BN modular exponentiation and RSA operations.

### Performance Improvement

```
openssl speed rsa2048
```

benchmark results run on SG2042:

| Test Item | Baseline (C) | ASM Optimized | Improvement |
| --------- | ------------ | ------------- | ----------- |
| ​sign/s    | 101.6        | 209.0         | 105.71%     |
| ​verify/s  | ​4004.0       | ​7270.1        | 81.57%      |
| encr./s   | 3804.2       | 6682.1        | ​75.65%      |
| decr./s   | 100.7        | 205.3         | ​103.87%     |

### Test Verification
All relevant tests pass in the default build configuration:
```bash
make test

All tests successful.
Files=345, Tests=4491, 2186 wallclock secs (32.44 usr  3.42 sys + 1719.95 cusr 772.27 csys = 2528.08 CPU)
Result: PASS
make[2]: Leaving directory '/home/mnt/fpb/submit/openssl-master'
make[1]: Leaving directory '/home/mnt/fpb/submit/openssl-master'
```
##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

t8m commented:
@ZenithalHourlyRate @HeliC829 @JerryShih  previously reviewed RISC-V assembly code.

ZenithalHourlyRate commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,446 @@
+    mulhu   $ahi, $aj, $m0       
+
+    mul     $m1, $lo0, $n0        
+    mv      sp, $tp               
```
if tp is (num + 2) * ULONG, I suggest allocaing tp at once instead of putting all these three lines apart

```asm
slli    $temp, $num, 3
sub     $tp, sp, $temp
andi(maybe addi)    $tp, $tp, -16
mv      sp, $tp
```

into
```asm
slli $num, $num, 3
sub sp, sp, $num
sub sp, sp, 16
mv  $tp, sp
```

fengpengboa replied:
I will make the modifications as suggested.

ZenithalHourlyRate commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,446 @@
+    addi    $tp, $tp, 8             
+    bnez    $j, .L1st           
+
+.L1st_skip:
```
I would name such section in a style like `1st_fix_tp_carry` instead of an informationless `skip`; same for `inner_skip` below.

ZenithalHourlyRate commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,446 @@
+    addi    $rp, $rp, 8
+    addi    $num,$num,-8
+    nop
+.Lcond_copy:
+    addi    $num,$num, -8         
+    bnez $carry1,.set_aj1 
+    mv   $nj,$tj
+    j    .end1
+.set_aj1:
+    mv   $nj,$aj         
```
These bnez/j are not constant-time. You can use `snez` to implement a constant time selection logic.

fengpengboa replied:
I will make the modifications as suggested

ZenithalHourlyRate reviewed (commented):
I have not reviewed all the logic, only covering some control logics here; the arithmetic part needs further reviewing; but I think some questions should be answered as they affect the arithmetic logic (i.e. the num >= 2 assumption)

The `andi -16` is confusing for me. If it is a typo for `addi`, how would `make test` pass?

Also, please add comments before each asm block (like when you separate them by an empty line so it is more readable; the variable names are a bit confusing, like `lo0,hi0,alo,ahi,lo1,hi1,aj,nj,i,j`, need some time to distinguish what 0/1 means and what `i/j` means.

fengpengboa commented:
> @fengpengboa Could you please get some independent review from somebody from the RISC-V community? We do not have RISC-V assembly experts among committers.

Thank you for the recommendation. Should the reviewer’s availability be limited, we will proactively identify qualified RISC-V experts for an independent review.

fengpengboa commented:
> I have not reviewed all the logic, only covering some control logics here; the arithmetic part needs further reviewing; but I think some questions should be answered as they affect the arithmetic logic (i.e. the num >= 2 assumption)
> 
> The `andi -16` is confusing for me. If it is a typo for `addi`, how would `make test` pass?
> 
> Also, please add comments before each asm block (like when you separate them by an empty line so it is more readable; the variable names are a bit confusing, like `lo0,hi0,alo,ahi,lo1,hi1,aj,nj,i,j`, need some time to distinguish what 0/1 means and what `i/j` means.

​​The instruction `andi $tp, $tp, -16` is used for memory address alignment, rather than an addi operation.In addition, I will supplement the relevant code comments, including the meaning of the variable names.

fengpengboa commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,446 @@
+
+$code .= <<___;
+
+.Lmul_mont:		
```
I might consider adding a branch here subsequently. If it's unnecessary now, I can delete it.

fengpengboa commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,446 @@
+.Lmul_mont:		
+    ld      $m0, 0($bp)         
+    addi    $bp, $bp, 8         
+    slli    $temp, $num, 3         
```
I will modify this part

fengpengboa commented:
@t8m  Hi, I've updated the code per review comments. Could you please approve the CI workflow? Thanks!

fengpengboa commented:
@ZenithalHourlyRate  Hello! I have now addressed all your previous review comments. The latest changes have been pushed. Would you please take a look ​​when you have a moment​​? If you have any further suggestions, feel free to let me know. Thank you for your guidance!

ZenithalHourlyRate reviewed (commented):
Generally LGTM! Just some questions and nitpicking comments.

paulidale commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,456 @@
+
+    mul $m1, $lo0, $n0    # montgomery reduction coefficient tp[0]*n0
+    # montgomery optimization: np[0]*m1 ensures (np[0]*m1+lo0) has zero lower bits
+    # only carry status needed ,not full lo1 result
```
NIT:
```suggestion
    # only carry status needed, not full lo1 result
```

fengpengboa commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,456 @@
+    mv $lo1, $temp
+    add $hi1, $hi1, $carry1
+
+    sd $lo1, 0($tp)    # tp[j-1]
+    addi $tp, $tp, 8
+
+    mul $alo, $aj, $m0    # ap[j]*bp[0]
+    mulhu $ahi, $aj, $m0
+    mul $nlo, $nj, $m1    # np[j]*m1
+    mulhu $nhi, $nj, $m1
```
> I understand that you want to compute `ap[j] * bp[0]` with j ranging from 2 to `num-1`, but `$j` here goes down from `num-2` to 0 so the `tp[j-1]` and `ap[j]` comments here may be hard to understand.
> 
> Also following the logic here, it should be `tp[j-2]` instead of `tp[j-1]`, due to montgomery reduction where zero lower bits have already been eliminated.
> 
> I would recommend a comment like
> 
> ```
> # $j ranges from (num-2)*8 to 0, however, the j in comment ranges from 2 to num
> ```

Thanks for the comment! I'll update the annotations per your suggestions to clarify the intent and check for similar cases.

fengpengboa commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,456 @@
+    sltu $carry2, $temp, $hi1
+    add $hi1, $temp, $carry1
+    sltu $ovf, $hi1, $temp
+    or $carry1, $carry2, $ovf
```
> Though I understand that only one of carry2 or ovf could be true, it is not so obvious so better add a comment here.

Thanks for the note! I will make the modifications as suggested.

fengpengboa commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,456 @@
+    sub $np, $np, $num    # rewind $np
+
+    # last reduction term
+    add $lo1, $nlo, $hi1    # {np[num-1]*m1,low}+{np[num-1]*m1,high}
```
> Update comment here.

Thanks for the note!

fengpengboa commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,456 @@
+    sub	$np, $np, $num    # rewind $np
+
+    # last reduction term
+    add $lo1, $nlo, $hi1    # {np[num-1]*m1,low}+{np[num-1]*m1,high}
```
> Update comment here.

Thanks for the note!

fengpengboa commented on crypto/bn/asm/riscv64-mont.pl:
```diff
@@ -0,0 +1,456 @@
+    # last partial product + last reduction term
+    add $lo0, $lo0, $tj
+    sltu $carry1, $lo0, $tj
+    add $hi0, $hi0, $carry1
```
> Is there guarantee that hi0 wont overflow? `hi0` here is `ahi + (carry of alo + hi0) + (carry of lo0 + tj)` and may overflow where `ahi` is high part of `ap[num-1]*bp[i]`. For addition of one carry it wont overflow (the (a*b+c) case) but for two carries it may overflow.

The overflow scenario you described is theoretically possible.If the value in hi0 is exactly 0xFFFFFFFFFFFFFFFF (all ones) and the previous addition operation generated a carry (C=1), then executing adc $hi0, $hi0, xzr could indeed cause the loss of the high-order carry bit. However, the probability of this specific condition occurring is extremely small.
Additionally, the existing merged ARM platform implementation follows the same design logic and also does not specifically handle this particular boundary condition.Interestingly, both the current ARM implementation and my submitted code pass all of OpenSSL's existing unit tests (including various edge case tests), and both implementations have shown no issues in practical testing so far.
Regarding why this problem hasn't manifested, I believe there are two possibilities:
1.Our current test suite may not yet cover this extremely rare edge scenario which has an exceedingly low probability of occurring. 2.In this specific context, the lost carry might not affect the final computation result.

fengpengboa commented:
Hi @t8m  I've implemented all review suggestions in the latest code changes. Would you be kind enough to re-approve the CI workflow when you have a moment? Really appreciate your guidance on this!

fengpengboa commented:
@ZenithalHourlyRate @paulidale I wanted to sincerely thank you all for taking the time to review the earlier version and provide your valuable insights. Following your guidance, I've now addressed all review comments in the latest code update. I'd be truly grateful if you could spare some time to examine the revised changes when your schedule permits.

fengpengboa commented:
> From my side I think it is ok.

Thank you for your review! I really appreciate the confirmation from your side.

fengpengboa commented:
@ZenithalHourlyRate Could you please add the formal approval checkmark when you have a moment?

fengpengboa commented:
@t8m @nhorman  Could you please help move this PR forward for merging? We have one approval and one verbal LGTM. Is there anything else required from our side?

fengpengboa commented:
@HeliC829 Hi, I would be grateful if you could find the time to review this PR. Thank you!

HeliC829 reviewed (commented):
LGTM.

fengpengboa commented:
> LGTM.

Thank you for the review!

fengpengboa commented:
Hi @InfoHunter  , could you please review this PR when you have time? Thank you!

paulidale commented:
Merged after removing end of line whitespace.  Thanks for the contribution.
